h5math.1
.\" Copyright (c) 1999-2009 Massachusetts Institute of Technology
.\" 
.\" Permission is hereby granted, free of charge, to any person obtaining
.\" a copy of this software and associated documentation files (the
.\" "Software"), to deal in the Software without restriction, including
.\" without limitation the rights to use, copy, modify, merge, publish,
.\" distribute, sublicense, and/or sell copies of the Software, and to
.\" permit persons to whom the Software is furnished to do so, subject to
.\" the following conditions:
.\" 
.\" The above copyright notice and this permission notice shall be
.\" included in all copies or substantial portions of the Software.
.\" 
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
.\" EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
.\" IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
.\" CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
.\" TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
.\" SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
.\"
.TH H5MATH 1 "May 23, 2005" "h5utils" "h5utils"
.SH NAME
h5math \- combine/create HDF5 files with math expressions
.SH SYNOPSIS
.B h5math
[\fIOPTION\fR]... \fIOUTPUT-HDF5FILE\fR [\fIINPUT-HDF5FILES\fR...]
.SH DESCRIPTION
.PP
." Add any additional description here
h5math takes any number of HDF5 files as input, along with a mathematical
expression, and combines them to produce a new HDF5 file.

HDF5 is a free, portable binary format and supporting library developed
by the National Center for Supercomputing Applications at the University
of Illinois in Urbana-Champaign.  A single
.I h5
file can contain multiple data sets; by default,
.I h5math
creates a dataset called "h5math", but this can be changed via the
.B -d
option, or by using the syntax \fIHDF5FILE:DATASET\fR.  The
.B -a
option can be used to append new datasets to an existing HDF5 file.
The same syntax is used to specify the dataset used in the input
file(s); by default, the first dataset (alphabetically) is used.

A simple example of h5math's usage is:
.IP "" 4
h5math -e "d1 + 2*d2" out.h5 foo.h5 bar.h5:blah
.PP
which produces a new file, out.h5, by adding the first dataset in
foo.h5 with twice the "blah" dataset in bar.h5.  In the expression
(specified by \fB-e\fR), the first input dataset (from left to right)
is referred to as \fId1\fR, the second as \fId2\fR, and so on.

In addition to input datasets, you can also use the x/y/z coordinates
of each point in the expression, referenced by "x" "y" and "z"
variables (for the first three dimensions) as well as a "t" variable
that refers to the last dimension.  By default, these are integers starting at 0 at the corner of the dataset, but the
.B -0
option will change the x/y/z origin to the center of the dataset (t is
unaffected), and the
.B -r
.I res
option will specify the "resolution", dividing the x/y/z coordinates
by \fIres\fR.

All of the input datasets must have the same dimensions, which are
also the dimensions of the output.  If there are no input files, and
you are defining the output purely by a mathematical formula, you can
specify the dimensions of the output explicitly via the
.B -n
.I size
option, where
.I size
is e.g. "2x2x2".

Sometimes, however, you want to use only a smaller-dimensional "slice"
of multi-dimensional data.  To do this, you specify coordinates in one
(or more) slice dimension(s), via the
.B -xyzt
options.
.SH OPTIONS
.TP
.B -h
Display help on the command-line options and usage.
.TP
.B -V
Print the version number and copyright info for h5math.
.TP
.B -v
Verbose output.
.TP
.B -a
If the HDF5 output file already exists, append the data as a new
dataset rather than overwriting the file (the default behavior).  An
existing dataset of the same name within the file is overwritten,
however.
.TP
\fB\-e\fR \fIexpression\fR
Specify the mathematical expression that is used to construct the
output (generally in " quotes to group the expression as one item in
the shell), in terms of the variables for the input datasets and
the coordinates as described above.

Expressions use a C-like infix notation, with most standard operators
and mathematical functions (+, sin, etc.) being supported.  This
functionality is provided (and its features determined) by GNU libmatheval.
.TP
\fB\-f\fR \fIfilename\fR
Name of a text file to read the expression from, if no
.B -e
expression is specified.  Defaults to stdin.
.TP
\fB\-x\fR \fIix\fR, \fB\-y\fR \fIiy\fR, \fB\-z\fR \fIiz\fR, \fB\-t\fR \fIit\fR
This tells
.I h5math
to use a particular slice of a multi-dimensional dataset.  e.g.
.B -x
uses the subset (with one less dimension) at an x index of
.I ix
(where the indices run from zero to one less than the maximum index in
that direction).  Here, x/y/z correspond to the first/second/third
dimensions of the HDF5 dataset. The \fB\-t\fR option specifies a slice
in the last dimension, whichever that might be.  See also the
.B -0
option to shift the origin of the x/y/z slice coordinates to the
dataset center.
.TP
.B -0
Shift the origin of the x/y/z slice coordinates to the dataset center,
so that e.g. -0 -x 0 (or more compactly -0x0) returns the central x
plane of the dataset instead of the edge x plane.  (\fB\-t\fR
coordinates are not affected.)

This also shifts the origin of the x/y/z variables in the expression so
that 0 is the center of the dataset.
.TP
\fB\-r\fR \fIres\fR
Use a resolution
.I res
for x/y/z (but not t) variables in the expression, so that the data
"grid" coordinates are divided by \fIres\fR.  The default \fIres\fR is 1.

For example, if the x dimension has 21 grid steps, setting a \fIres\fR
of 20 will mean that x variables in the expression run from 0.0 to
1.0 (or -0.5 to 0.5 if \fB\-0\fR is specified), instead of 0 to 20.

.B -r
does not affect the coordinates used for slices, which are always integers.
.TP
\fB\-n\fR \fIsize\fR
The output dataset must be the same size as the input datasets.  If
there are no input datasets (if you are defining the output purely by
a formula), then you must specify the output size manually with this
option: \fIsize\fR is of the form MxNxLx... (with M, N, L being
integers) and may be of any dimensionality.
.TP
\fB\-d\fR \fIname\fR
Write to dataset
.I name
in the output; otherwise, the output dataset is called "data" by default.
Also use dataset
.I name
in the input; otherwise, the first input dataset (alphabetically) in a
file is used.  Alternatively, use the syntax \fIHDF5FILE:DATASET\fR
(which overrides the
.B -d
option).
.SH BUGS
Send bug reports to S. G. Johnson, stevenj@alum.mit.edu.
.SH AUTHORS
Written by Steven G. Johnson.  Copyright (c) 2005 by the Massachusetts
Institute of Technology.


NEWS
h5utils 1.12 (6/12/09)

  * The vertical axis in h5topng is now reversed to correspond to
    what most people seem to expect: increasing coordinates correspond
    to "up" and "right" in the image, rather than "down" and "right"
    in the image as in previous versions.

  * Fixed failure in h5tovtk -2; thanks to Karen Lee for the bug report.

  * Fixed installation of h5read.oct for Octave 3.x.

h5utils 1.11.1 (4/28/08)

  * Fixed failure to find colormap files in h5topng 1.11 (due
    to changes in autoconf 2.60); thanks to bug report from Jiangjun Zheng.

h5utils 1.11 (4/24/08)

  * h5tovtk no longer reverses the dimensions; thanks to Andreas Wilde
    for the suggestion.

  * Fix compilation failure with HDF5 1.8.

h5utils 1.10.1 (9/20/06)

  * Fixed build problem on Cygwin due to missing ".exe" extension. 
    Thanks to Ken Hill for the bug report.

h5utils 1.10 (9/2/05)

  * Added h4fromh5 utility.  (NCSA seems to be no longer shipping
    the HDF5/HDF4 conversion tools with the latest HDF5 release.)

  * Added dkbluered color map, which is similar to bluered but
    uses a somewhat wider range of colors.

h5utils 1.9.1 (8/5/04)

  * Fix h5topng compilation failure with some non-C99 compilers;
    thanks to Maarten van Reeuwijk for the bug report.

h5utils 1.9 (7/12/04)

  * Added new h5math utility, which creates and combines HDF5 datasets
    using a user-specified mathematical expression.  (Requires GNU
    libmatheval to be installed.)

  * h5topng: Matlab-like start:end or start:step:end notation for
    slice indices, to allow a whole sequence of slices to be output as
    a sequence of PNG images.

  * h5topng: if contour/overlay dataset does not have same dimensions
    as output data, it is periodically "tiled" over the output.

h5utils 1.8 (5/22/04)

  * New -A and -a options for h5topng to allow translucent overlays
    from one file onto another (an alternative or complement to the
    -C contour-overlay option).

  * h5topng uses 24-bit direct color by default (use -8 option for
    old 8-bit behavior).

  * h5topng uses columns/rows for x/y by default (use -T to swap),
    the opposite of the old behavior.

  * There is no default -z 0 slice dimension in h5topng/h5totxt any more.
    You must specify a slice for 3+ dimensional data in h5topng.
    h5totxt dumps the whole data file by default unless one or more
    slices are specified.

  * Support specifying multiple slice dimensions for 4+ dimensional
    datasets, with new "-t" option to indicate final dimension.

  * Slices are also supported now in h5tovtk and h5tov5d.

  * New -. option in h5totxt to specify number of significant digits; 
    output 16 digits by default instead of 6, previously.

h5utils 1.7.1 (3/16/02)

  * Fixed array overrun in h5topng that caused a floating-point exception
    on Alphas; thanks to Marin Soljacic for the bug report.

h5utils 1.7 (3/9/02)

  * h5topng now supports multiple, user-definable color tables, a number
    of which are provided.  INCOMPATIBLE CHANGE: the -c option now has
    the syntax: -c <colortable>.  The old behavior corresponds to the
    included "bluered" colortable, invoked via: -c bluered.

  * New -R option for h5topng to use a consistent color scale for
    all specified files.

  * New -0 option for h5topng and h5totxt that shifts the origin of the
    slice coordinates to the dataset center.

  * Added h5tovtk program to output VTK (Visualization ToolKit) data files.

  * Support -T (transpose dimensions) option in h5tov5d.

  * Fixed bug in h5topng that caused extra rows/columns of garbage pixels
    to be written at the edges of images when scaling was used.

  * When compiling the h5read Octave plugin, respect the CPPFLAGS and LDFLAGS 
    environment variables.  Thanks to Max Colice for the bug report.

  * Fixed problem when --without-h5tov5d and --without-h5fromh4 are used. 
    Thanks to Nikola Ivanov Nikolov for the fix.

h5utils 1.6 (1/17/01)

  * Don't build h5fromh4 if the superior h4toh5 tool (from HDF5 1.4)
    is present.  Also added --{with,without}-h5fromh4 option to
    configure to force whether h5fromh4 is built.

h5utils 1.5.1 (12/9/00)

  * Support manually disabling Octave plugin support
    (configure --without-octave) in case of C++ problems.

  * Support Vis5d+ and Debian HDF header file locations.

h5utils 1.5 (7/9/00)

  * Added h5fromh4 program to convert HDF4 datasets to HDF5 format.

  * Added -S <s> option to h5topng as a shortcut for -X <s> -Y <s>.

h5utils 1.4 (5/28/00)

  * Added h5fromtxt program to convert text input to an HDF5 dataset.

h5utils 1.3.4 (1/31/00)

  * Improved -C contour plotting in h5topng.

  * Fix in h5topng man page (thanks to Christoph Becher).

h5utils 1.3.3 (1/30/00)

  * Bug fix in h5topng (would sometimes output solid black images).
    Thanks to Karl Koch for the bug report.

h5utils 1.3.2 (1/28/00)

  * Added h5topng -Z option to center color scale on zero.

  * Now support h5topng -C filename:dataset.

h5utils 1.3.1 (1/27/00)

  * Bug fixes in dataset name-handling, especially when using h5tov5d to
    join multiple datasets into one output file.

h5utils 1.3 (1/21/00)

  * You can now specify individual datasets within a file by using
    <filename>:<dataset> instead of just <filename> with h5topng,
    h5totxt, and h5tov5d.

h5utils 1.2.3: (1/20/00)

  * Fixed minor bug in 'h5totxt -h'.

h5utils 1.2.2: (1/12/00)

  * Makefile now includes CPPFLAGS in the compiler flags, making it
    easier to use header files in non-standard locations.  (CPPFLAGS
    is the proper place to put -I flags for the configure script.)

h5utils 1.2.1:  (1/1/00)

  * Modified -o option of h5tov5d to join datasets into a single Vis5d file.

h5utils 1.2:  (12/31/99)

  * Added h5tov5d program for converting to Vis5d format.

  * Improved installation; h5read.oct now goes into the sitewide Octave
    plugins directory, and things work correctly when only some of the
    utilities are compiled.


h5math.c
/* Copyright (c) 1999-2009 Massachusetts Institute of Technology
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 * 
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#include <unistd.h>

#include "config.h"
#include "arrayh5.h"
#include "copyright.h"
#include "h5utils.h"

#include <matheval.h>

#define CHECK(cond, msg) { if (!(cond)) { fprintf(stderr, "h5math error: %s\n", msg); exit(EXIT_FAILURE); } }

const char default_data_name[] = "h5math";

void usage(FILE *f)
{
     fprintf(f, "Usage: h5math [options] <output-filename> [<filenames>]\n"
	     "Options:\n"
	     "         -h : this help message\n"
             "         -V : print version number and copyright\n"
	     "         -v : verbose output\n"
	     "         -a : append to existing hdf5 file\n"
	     "  -n <size> : output array dimensions [ default: from input ]\n"
	     "  -f <file> : read expression to evaluate from file [ default: stdin ]\n"
	     "  -e <expr> : evaluate <expr> to output\n"
	     "    -x <ix> : take x=<ix> slice of data\n"
	     "    -y <iy> : take y=<iy> slice of data\n"
	     "    -z <iz> : take z=<iz> slice of data\n"
	     "    -t <it> : take t=<it> slice of data's last dimension\n"
	     "         -0 : use dataset center as origin for -x/-y/-z\n"
	     "     -r <r> : use resolution <r> for xyz coordinate units in expression\n"
	     "  -d <name> : use dataset <name> in the input/output files\n"
	     "              [ default: first dataset/%s ]\n"
	     "              -- you can also specify a dataset via <filename>:<name>\n",
	     default_data_name
	  );
}

#define MAX_RANK 10

int main(int argc, char **argv)
{
     arrayh5 *a, ao;
     int i, n;
     int rank = -1, dims[MAX_RANK];
     extern char *optarg;
     extern int optind;
     int c;
     int slicedim[4] = {NO_SLICE_DIM,NO_SLICE_DIM,NO_SLICE_DIM,NO_SLICE_DIM};
     int islice[4], center_slice[4] = {0,0,0,0};
     int verbose = 0;
     int append = 0;
     char *expr_string = 0, *expr_filename = 0;
     char *data_name = 0;
     char *out_fname, *out_dname;
     char **eval_vars;
     int eval_nvars;
     char **vars;
     double *vals;
     void *evaluator;
     double res = 1.0;
     int nx, ny, nz, nt, nr, ix, iy, iz, it, ir;
     double cx, cy, cz;

     while ((c = getopt(argc, argv, "hVvan:f:e:x:y:z:t:0d:r:")) != -1)
	  switch (c) {
	      case 'h':
		   usage(stdout);
		   return EXIT_SUCCESS;
	      case 'V':
		   printf("h5totxt " PACKAGE_VERSION " by Steven G. Johnson\n" 
			  COPYRIGHT);
		   return EXIT_SUCCESS;
	      case 'v':
		   verbose = 1;
		   break;
	      case 'a':
		   append = 1;
		   break;
	      case 'n':
	      {
		   int pos = 0;
		   rank = 0;
		   while (isdigit(optarg[pos])) {
			CHECK(rank < MAX_RANK,
			      "Rank too big in -n argument!\n");
			dims[rank] = 0;
			while (isdigit(optarg[pos])) {
			     dims[rank] = dims[rank]*10 + optarg[pos]-'0';
			     ++pos;
			}
			++rank;
			if (optarg[pos] == 'x' || optarg[pos] == 'X'
			    || optarg[pos] == '*')
			     ++pos;
		   }
		   CHECK(rank > 0 && !optarg[pos], "Invalid -n argument; should be e.g. 23x34 or 10x10x10\n");
		   break;
	      }
	      case 'f':
		   free(expr_filename);
		   expr_filename = my_strdup(optarg);
		   break;
	      case 'e':
		   free(expr_string);
		   expr_string = my_strdup(optarg);
		   break;
	      case 'x':
		   islice[0] = atoi(optarg);
		   slicedim[0] = 0;
		   break;
	      case 'y':
		   islice[1] = atoi(optarg);
		   slicedim[1] = 1;
		   break;
	      case 'z':
		   islice[2] = atoi(optarg);
		   slicedim[2] = 2;
		   break;
	      case 't':
		   islice[3] = atoi(optarg);
		   slicedim[3] = LAST_SLICE_DIM;
		   break;
              case '0':
                   center_slice[0] = center_slice[1] = center_slice[2] = 1;
                   break;
	      case 'r':
		   res = atof(optarg);
		   break;
	      case 'd':
		   free(data_name);
		   data_name = my_strdup(optarg);
		   break;		   
	      default:
		   fprintf(stderr, "Invalid argument -%c\n", c);
		   usage(stderr);
		   return EXIT_FAILURE;
	  }
     if (optind == argc) {  /* no parameters left */
	  usage(stderr);
	  return EXIT_FAILURE;
     }

     out_fname = split_fname(argv[optind], &out_dname);
     if (!out_dname[0]) {
	  if (data_name)
	       out_dname = data_name;
	  else
	       out_dname = (char *) default_data_name;
     }
     optind++;

     n = argc - optind;
     a = (arrayh5 *) malloc(sizeof(arrayh5) * n);
     CHECK(a, "out of memory");

     for (i = 0; i < n; ++i) {
	  int err;
	  char *fname, *dname;

          fname = split_fname(argv[i + optind], &dname);
          if (!dname[0])
               dname = data_name;

	  err = arrayh5_read(&a[i], fname, dname, NULL,
                             4, slicedim, islice, center_slice);
          CHECK(!err, arrayh5_read_strerror[err]);

	  CHECK(!i || arrayh5_conformant(a[i], a[i-1]),
		"all input arrays must have the same dimensions");

	  if (verbose)
	       printf("read variable d%d: dataset \"%s\" in file \"%s\"\n",
		      i + 1, dname ? dname : "<first>", fname);
	  
	  free(fname);
     }

     if (rank >= 0) {
	  ao = arrayh5_create(rank, dims);
	  CHECK(!n || arrayh5_conformant(ao, a[0]),
		"-n dimensions must be same as those of input arrays");
     }
     else if (n)
	  ao = arrayh5_clone(a[0]);
     else
	  CHECK(0, "output size must be specified with -n if no input arrays");

     if (verbose) {
	  printf("rank-%d array dimensions: ", ao.rank);
	  if (!ao.rank) printf("1\n");
	  for (i = 0; i < ao.rank; ++i)
	       printf("%s%d", i ? "x" : "", ao.dims[i]);
	  printf("\n");
     }

     nx = ao.rank >= 1 ? ao.dims[0] : 1;
     ny = ao.rank >= 2 ? ao.dims[1] : 1;
     nz = ao.rank >= 3 ? ao.dims[2] : 1;
     nt = ao.rank >= 4 ? ao.dims[3] : 1;
     for (nr = 1, i = 4; i < ao.rank; ++i)
	  nr *= ao.dims[i];
     cx = center_slice[0] ? (nx - 1) * 0.5 : 0.0;
     cy = center_slice[1] ? (ny - 1) * 0.5 : 0.0;
     cz = center_slice[2] ? (nz - 1) * 0.5 : 0.0;

     vars = (char **) malloc(sizeof(char *) * (n + 4));
     CHECK(vars, "out of memory");
     vals = (double *) malloc(sizeof(double) * (n + 4));
     CHECK(vals, "out of memory");
     for (i = 0; i < n; ++i) {
	  vars[i] = my_strdup("dxxxxxxxxxxxx");
#ifdef HAVE_SNPRINTF
	  snprintf(vars[i], 14, "d%d", i + 1);
#else
	  sprintf(vars[i], "d%d", i + 1);
#endif
	  vals[i] = 0.0;
     }
     vars[n+0] = strdup("x"); vals[n+0] = 0.0;
     vars[n+1] = strdup("y"); vals[n+1] = 0.0;
     vars[n+2] = strdup("z"); vals[n+2] = 0.0;
     vars[n+3] = strdup("t"); vals[n+3] = 0.0;
     
     if (!expr_string) {
	  char buf[1024] = "";
	  int len;
	  FILE *f = expr_filename ? fopen(expr_filename, "r") : stdin;
	  CHECK(f, "unable to open expression file");

	  if (verbose && f == stdin)
	       printf("Enter expression to write to %s:\n", out_fname);
	  
	  fgets(buf, 1024, f);
	  expr_string = my_strdup(buf);
	  len = strlen(buf) + 1;
	  while (fgets(buf, 1024, f)) {
	       len += strlen(buf);
	       expr_string = (char *) realloc(expr_string, len);
	       strcat(expr_string, buf);
	  }
	  for (ix = 0; ix < len; ++ix)
	       if (expr_string[ix] == '\n')
		    expr_string[ix] = ' '; /* matheval chokes on newlines */

	  if (expr_filename) fclose(f);
     }

     CHECK(evaluator = evaluator_create(expr_string),
	   "error parsing symbolic expression");

     evaluator_get_variables(evaluator, &eval_vars, &eval_nvars);
     for (ix = 0; ix < eval_nvars; ++ix) {
	  for (iy = 0; iy < n + 4 && strcmp(eval_vars[ix], vars[iy]); ++iy)
	       ;
	  if (iy == n + 4) {
	       fprintf(stderr, "h5math error: unrecognized variable \"%s\"\n",
		       eval_vars[ix]);
	       exit(EXIT_FAILURE);
	  }
     }
     
     if (verbose) {
	  char *buf = evaluator_get_string(evaluator);
	  printf("Evaluating expression: %s\n", buf);
     }

     for (ix = 0; ix < nx; ++ix)
     for (iy = 0; iy < ny; ++iy)
     for (iz = 0; iz < nz; ++iz)
     for (it = 0; it < nt; ++it)
     for (ir = 0; ir < nr; ++ir) {
	  int idx = ir + nr * (it + nt * (iz + nz * (iy + ny * ix)));
	  for (i = 0; i < n; ++i)
	       vals[i] = a[i].data[idx];
	  vals[n+0] = (ix - cx) / res;
	  vals[n+1] = (iy - cy) / res;
	  vals[n+2] = (iz - cz) / res;
	  vals[n+3] = ao.rank >= 4 ? it : 
	       (ao.rank >= 3 ? iz : (ao.rank >= 2 ? iy : ix));
	  ao.data[idx] = evaluator_evaluate(evaluator, n+4, vars, vals);
     }

     if (verbose)
	  printf("Writing data to \"%s\" in \"%s\"...\n", 
		 out_dname ? out_dname : "<first>", out_fname);
     arrayh5_write(ao, out_fname, out_dname, append);

     free(vals);
     for (i = 0; i < n+4; ++i) free(vars[i]);
     free(vars);
     arrayh5_destroy(ao);
     for (i = 0; i < n; ++i) arrayh5_destroy(a[i]);
     free(a);
     free(out_fname);
     free(expr_filename);
     free(expr_string);
     free(data_name);

     return EXIT_SUCCESS;
}
